Source/OxyPlot/Series/ContourSeries.cs

```diff
@@ -486,100 +459,118 @@ private void AddContourLabels(Contour contour, ScreenPoint[] pts, ICollection<Co
+                    // as points are sorted by X, we typically only need to check the point immediately following the current point
+                    do
                     {
-                        contourPoints.Insert(0, reverse ? segment1.StartPoint : segment1.EndPoint);
-                        contourPointsCount++;
-                        this.segments.Remove(segment1);
-                        segmentCount--;
+                        if (i + i2 >= points.Count)
+                        {
+                            joinPoint = null;
+                            break;
+                        }
+
+                        joinPoint = points[i + i2];
+                        i2++;
+                        if (joinPoint.Join != null)
+                        {
+                            continue;
+                        }
+
+                        if (joinPoint.Point.X > maxX)
+                        {
+                            joinPoint = null;
+                            break;
+                        }
                     }
+                    while (Math.Abs(joinPoint.Point.Y - currentPoint.Point.Y) > eps);
 
-                    // Find a segment that is connected to the tail of the contour
-                    segment2 = this.FindConnectedSegment(contourPoints[contourPointsCount - 1], firstSegment.ContourLevel, eps, out reverse);
-                    if (segment2 != null)
+                    // join the two points together
+                    if (joinPoint != null)
                     {
-                        contourPoints.Add(reverse ? segment2.StartPoint : segment2.EndPoint);
-                        contourPointsCount++;
-                        this.segments.Remove(segment2);
-                        segmentCount--;
+                        currentPoint.Join = joinPoint;
+                        joinPoint.Join = currentPoint;
                     }
                 }
 
-                if ((segment1 == null && segment2 == null) || segmentCount == 0)
+                // go through the points again, this time we follow the joined point chains to obtain the contours
+                foreach (var segmentPoint in points)
                 {
-                    if (contourPointsCount > 0 && firstSegment != null)
+                    if (segmentPoint.Processed)
+                    {
+                        continue;
+                    }
+
+                    var currentPoint = segmentPoint;
+                    var dataPoints = new List<DataPoint>();
+
+                    if (currentPoint.Join == null) // we are the first point of the contour
+                    {
+                        dataPoints.Add(currentPoint.Point);
+                        currentPoint = currentPoint.Partner;
+                        dataPoints.Add(currentPoint.Point);
+                    }
+                    else // we are somewhere in the middle of the contour, our partner might be the first, so add it first
```

Comment:
Not sure I understand this bit

Reply:
Here we have two possibilities: 
1. The `currentPoint` is _not_ joined to any other point -> This will be the case for e.g. the first point of an open contour. In this case `currentPoint` will have to be the start point of our contour.

2. The `currentPoint` _is_ joined to another point -> This means that:
2.1 **Either** the contour is closed, so it doesn't matter if we add the `currentPoint` or its `Partner` first
2.2 **Or** the contour is not closed. In that case the `currentPoint`'s `Partner` _may_ be the start point of the contour -> in this case we would have to add the `currentPoint`'s `Partner` first

So in the second case, we just always add the `currentPoint`'s `Partner` first to be safe.

I do realize that this sounds really convoluted.. in my mind this made perfect sense earlier. I'll take another look tomorrow with fresh eyes and see if I can simplify this a bit. Or maybe the naming just isn't right, but I couldn't think of better alternatives to `Partner` and `Join`...

Reply:
`Partner` and `Join` are OK in my opinion. What I'm missing is the reason why it can't be the case that both `currentPoint` and `currentPoint.Partner` could be 'in the middle' of an open contour. The partner _might_ be first, but is that necessarily the case?

Reply:
I flipped `Peaks` around, so that there is an 'open' contour which ends with X large, and you can see the effect of this when you plot the start/end of each contour:

![image](https://user-images.githubusercontent.com/5646475/96378138-014a4f80-1182-11eb-99d0-7f6e62bc1cd1.png)

I think the best thing to do is to follow the contour to the end: if `currentPoint` gets back to `segmentPoint`, then you know it is a closed loop and, and can start anywhere; if you get to an end (i.e. a point that isn't joined to another point) then you start from the end.

Reply:
Oh yes, of course you are right, I'll have to fix that.

Reply:
Should be fixed now. I also slightly changed the `peaks` function used in the example to have all kinds of open contours included in the example.

Reply:
Oh no! You can't change peaks! It's the logo!

Reply:
New implementation looks good to me, and the issues seems to be fixed. I don't know if we do care about changing `peaks`, but the new one also doesn't cover the case of the algorithm finding the middle of an open contour initially: the open loops in the new example all 'start' at a low X value.

Reply:
Haha, I'll try to come up with a function that will also cover contours which have their lowest X value somewhere in the middle and make that a separate example, so we can keep the original `peaks` :)